src/api/events.ts

```diff
@@ -53,3 +53,60 @@ export const venueEventWithIdConverter: firebase.firestore.FirestoreDataConverte
     return withId(snapshot.data() as VenueEvent, snapshot.id);
   },
 };
+
+export interface ImportEventsBatchError {
+  event: VenueEvent;
+  index: number;
+  reason: string;
+  venueId: string;
+}
+
+export interface ImportEventsBatchResult {
+  errors: ImportEventsBatchError[];
+  updateCount: number;
+}
+
+export const importEventsBatch: (
+  data: VenueEventWithVenueId[]
+) => Promise<ImportEventsBatchResult> = async (data) => {
+  if (!data) {
+    return {
+      errors: [],
+      updateCount: 0,
+    };
+  }
+
+  const venuesRef = firebase.firestore().collection("venues");
```

Comment:
> This will load all of the venues into memory, which is not good.

Unqualified like that, I can't agree. In most cases probably not good, in few - probably unavoidable.

The issue I see is a bit different:

Is this solution adequate enough for a rare event like importing CSV file in Admin by select few people where the general audience has no access to and how much would that impact justify adding new functionality that will probably be OOS to this PR?

On the plus side, the actual venues data (all their fields and sub-collections) isn't being read (no `.data()` nor `.get()` used):

> A CollectionReference object can be used for adding documents, getting document references, and querying for documents (using the methods inherited from Query).

> A QueryDocumentSnapshot contains data read from a document in your Firestore database as part of a query. The document is guaranteed to exist and its data can be extracted with .data() or .get(<field>) to get a specific field.

Reply:
You shouldn't fetch all of the venues just to check if your venues exist in the database. You can either create an array of your venues and use promise all to check if they are in the database, or just fetch only the related venues instead of all of them. But fetching the whole collection is definitely not the correct way to do it.

We had a client with 3000 venues. If your CSV has 2 venue ids, this request will fetch 3000 venues which is going to take I don't know how long, and then it will have to iterate over these 3000 venues. All this just for a simple validation.

Reply:
The validation is probably no longer necessary. Now there is a check on update (can't remember if cloud function or just the DB rules) that only venues the user is listed as owner are viable to be updated. That should probably solve the trouble of fetching and double-checking